all/CMakeLists.txt

```diff
@@ -0,0 +1,57 @@
+# This file should be used for development with IDEs for the best discovery
```

Comment:
The CML in the `all` directory is a superbuild CML, while the one in the root and the test directory are project CMLs. The difference is that project CMLs should strictly only with build and usage requirements, while superbuild CMLs can serve as a containment for the gruesome details, like:

* Header only libraries don't have build artifacts, which makes it impossible to properly edit them from all IDEs, VS being one example. Creating a dummy target that does build an artifact makes VS discover the sources and you can edit the project normally.
* Disabling the warning guard in CI.
* Add a dummy config file so the `find_package(Result)` call in the test project CML becomes a noop, as the targets are already present due to the library being `add_subdirectory`'d beforehand.
* Fix Catch2's broken conan package :)

Reply:
I'm starting to see what you're getting at with this, but this strikes me as extremely subjective and opinionated approach.

For starters, consumers of a library rarely need to grok that library's CMakeLists.txt unless they really care for custom options or specifics. This means it primarily benefits the creator and/or maintainers, of which there's little real benefit outside of your point about "hiding the gruesome details", which really there aren't many of.

As far as I can tell, the only differences between the two approaches are:

* You've moved some configuration to a different CMakeLists.txt, and
* Your approach requires configuring `all/CMakeLists.txt`, mine requires configuring just `CMakeLists.txt` with a single option.

At best this seems like a lateral trade to me, but with no clear benefit -- especially since lots of the configuration could be moved around just the same so that they're not seen as much by consumers

> Header only libraries don't have build artifacts, which makes it impossible to properly edit them from all IDEs, VS being one example. Creating a dummy target that does build an artifact makes VS discover the sources and you can edit the project normally.

This is already done with the self-containment test. Simply enabling this during configuring is all that's needed for this to work (e.g. the one option).

> Disabling the warning guard in CI.

Covered by being hidden; but really this is tiny. To be honest, these could actually just be moved to the `test/CMakeLists.txt` file anyway -- because it only really applies there. These flags aren't exposed to consumers (hence `add_compile_flags` and not `target_add_compile_flags( ... INTERFACE ...)` ) because most users don't care for warning flags. They're mostly here to ensure that consumers don't accidentally see undesirable warnings when using this code.

> Add a dummy config file so the `find_package(Result)` call in the test project CML becomes a noop, as the targets are already present due to the library being add_subdirectory'd beforehand.

I feel like I'm missing why this is a good thing. This sounds like a hack necessary to get this approach to work.

> Fix Catch2's broken conan package :)

Not broken (see my comment about missing the `conanbuildinfo.cmake`)

-------------------------

I appreciate you taking the time to show me this approach, as it is something new to me. However I don't really see any real benefits -- at least for this specific project.

Reply:
> extremely subjective and opinionated approach

You are absolutely correct :)

> consumers of a library rarely need to grok that library's CMakeLists.txt

This is a tricky topic. I have never seen a project write their CML in a way that lends itself to be easily understood, so the time investment of trying to understand someone's spaghetti code is indeed too high. But what if people started to exercise the same disciplines for CMake that they use for their "actual" code? A lot of the spaghetti is really unnecessary and project CMLs would be much more readable. The CMLs I strive to write are just a long list of requirements with little to no branching, trying not to use commands other than CMake provided ones as best as I can (sometimes it makes sense to abstract away a lot of properties into a macro that is applied to a lot of targets, there arguments on both sides), while constraining them into logically coherent blocks that I separate with those `# ----` comments.

> At best this seems like a lateral trade to me, but with no clear benefit

The benefit is that ExternalProject, FetchContent, CPM.cmake, vcpkg and `add_subdirectory` all expect a single CML at the root that corresponds to the project, in your case the library. In vcpkg's case you are putting a burden on the package maintainers to figure out how your project should be built and patch their way around things. As I mentioned elsewhere, these methods are becoming more and more popular to consume other projects with, and writing CMLs that are easy to consume in these ways are easy.  
In a perfect world (but even in an imprefect one like ours) we should be able to just use the command `cmake -S . -B build` to get a build tree with sane defaults that are still highly configurable to meet someone's special needs. I don't think I have to say how up and coming languages and technologies change their defaults to far more sane ones than the ones before them much to everyone's liking.

> these could actually just be moved to the test/CMakeLists.txt file anyway

I try to go for an approach where the test subproject pretends to be an independent project that just so happens to consume the library as a dependency, which means that you are consuming the library the same way everyone else would (and I think I mentioned elsewhere that installing ought to be the default mode on consumption). Placing this kind of constraint on myself, I think this results in more coherent CMLs.

> `add_compile_options`

This command and friends are traditional CMake, please don't use them.

>This sounds like a hack necessary to get this approach to work.

Not really a hack. In the superbuild CML you have complete knowledge of the entire project, the targets are already available and the package is not installed anywhere, so stubbing from the superbuild only makes sense. I'm not alone on this, you should watch [Daniel Pfeifer's CMake talk](https://www.youtube.com/watch?v=bsXLMQ6WgIk) where he mentions this exact same thing, but he probably didn't know that `find_*` commands are idempotent, as in they store the file they found in a cache variable and subsequent runs just rerun/return the same file they found before.

> Not broken

Yes, it is. See the doctest example above.